express-notes-api/index.js

```diff
@@ -0,0 +1,106 @@
+const express = require('express');
+const app = express();
+const notebookData = require('./data.json');
+const fs = require('fs');
+
+const notebook = notebookData.notes;
+const noteId = notebookData.nextId;
+
+app.use(express.json());
+
+app.get('/api/notes', (req, res) => {
+  const notesArr = [];
+  for (const key in notebook) {
+    notesArr.push(notebook[key]);
+  }
+  res.json(notesArr);
+});
+
+app.get('/api/notes/:id', (req, res) => {
+  const id = req.params.id;
+  if (id < 1 || isNaN(id)) {
+    const error = { error: 'id must be a positive integer' };
+    res.status(400).send(error);
+  } else if (!notebook[id]) {
+    const error = { error: `cannot find note with ${id}` };
+    res.status(404).send(error);
+  } else {
+    res.status(200).json(notebook[id]);
+  }
+});
+
+app.post('/api/notes', (req, res) => {
+  if (!req.body.content) {
+    const error = { error: 'content is a required field' };
+    res.status(400).send(error);
+  } else {
+    const newNote = {
+      id: noteId,
+      content: req.body.content
+    };
+    notebookData.notes[noteId] = newNote;
+    notebookData.nextId++;
+    const notebookJSON = JSON.stringify(notebookData, null, 2);
+    fs.writeFile('data.json', notebookJSON, 'utf8', err => {
+      if (err) {
+        const error = { error: 'An unexpected error occurred' };
+        res.status(500).send(error);
+      }
+    });
+    res.status(201).send(newNote);
```

Comment:
Be careful not to respond until the writeFile completes. Check your other uses of writeFile too.

Reply:
Hey Tim, I just want clarification on this comment!
`res.status(201).send(newNote);`
If this line is outside of the writeFile code block, shouldn't the response be after the writeFile function completes?

Reply:
@paulpark330 The only way to make sure that something happens after an asynchronous operation is to put it in the callback's code block.

Reply:
Definitely forgot about the asynchronous part of the writeFile method.
I will make more adjustments!
Thank you